rpc_new/client.ts

```diff
@@ -0,0 +1,169 @@
+import { Deferred, deferred } from "../deps/std/async.ts";
+import { getOrInit } from "../util/mod.ts";
+import * as msg from "./messages.ts";
+import { Provider, ProviderListener } from "./provider/base.ts";
+import { ProviderHandlerError, ProviderSendError } from "./provider/errors.ts";
+import * as U from "./util.ts";
+
+// TODO: delete this upon solving inner-type-access problem of RPC effects
+export declare const ClientE_: unique symbol;
+export type ClientE_ = typeof ClientE_;
+
+export class Client<
+  DiscoveryValue = any,
+  SendErrorData = any,
+  HandlerErrorData = any,
+  CloseErrorData = any,
+> {
+  // TODO: delete this as well pending the above `TODO` ^
+  declare [ClientE_]: {
+    send: SendErrorData;
+    handler: HandlerErrorData;
+    close: CloseErrorData;
+  };
+
+  providerRef;
+  pendingCalls: Record<string, Deferred<unknown>> = {};
+  pendingSubscriptions: SubscriptionListeners<SendErrorData, HandlerErrorData> = {};
+  activeSubscriptions: SubscriptionListeners<SendErrorData, HandlerErrorData> = {};
+  activeSubscriptionByMessageId: Record<string, string> = {};
+  subscriptionStates = new Map<string, WeakMap<new() => any, any>>();
+
+  constructor(
+    readonly provider: Provider<DiscoveryValue, SendErrorData, HandlerErrorData, CloseErrorData>,
+    readonly discoveryValue: DiscoveryValue,
+  ) {
+    this.providerRef = provider(discoveryValue, this.#listener);
+  }
+
+  #listener: ProviderListener<SendErrorData, HandlerErrorData> = (e) => {
+    if (e instanceof Error) {
+      for (const id in this.pendingCalls) {
+        const pendingCall = this.pendingCalls[id]!;
+        pendingCall.resolve(e);
+        delete this.pendingCalls[id];
+        this.pendingSubscriptions[id]!(e);
+        delete this.pendingSubscriptions[id];
+      }
+      for (const id in this.activeSubscriptions) {
+        this.activeSubscriptions[id]!(e);
+        delete this.activeSubscriptions[id];
+        this.subscriptionStates.delete(id);
+      }
```

Comment:
Seems pretty nuclear. Do errors never correspond to particular calls / subscriptions?

Reply:
Not send/handler/close errors. If we encounter those, there's an internal error and it's likely best that we throw away the provider instance and restart––we don't currently tackle this recovery, but we certainly will. Thoughts @kratico?

Reply:
right now, if there is a network error we don't do any recovery.

Errors do not always correspond to particular calls, you could have 2 active subscriptions and get the WebSocket closed.

once this PR is merged we can re-visit connection/disconnection recovery.